vault/resource_approle_auth_backend_role.go

```diff
@@ -0,0 +1,384 @@
+			},
+			"secret_id_ttl": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Description: "Number of seconds a SecretID remains valid for.",
+			},
+			"token_num_uses": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Description: "Number of times issued tokens can be used. Setting this to 0 or leaving it unset means unlimited uses.",
+			},
+			"token_ttl": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Description: "Default number of seconds to set as the TTL for issued tokens and at renewal time.",
+			},
+			"token_max_ttl": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Description: "Number of seconds after which issued tokens can no longer be renewed.",
+			},
+			"period": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Description: "Number of seconds to set the TTL to for issued tokens upon renewal. Makes the token a periodic token, which will never expire as long as it is renewed before the TTL each period.",
+			},
+			"backend": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Unique name of the auth backend to configure.",
+				ForceNew:    true,
+				Default:     "approle",
+				// standardise on no beginning or trailing slashes
+				StateFunc: func(v interface{}) string {
+					return strings.Trim(v.(string), "/")
+				},
+			},
+		},
+	}
+}
+
+func approleAuthBackendRoleCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	backend := d.Get("backend").(string)
+	role := d.Get("role_name").(string)
+
+	path := approleAuthBackendRolePath(backend, role)
+
+	log.Printf("[DEBUG] Writing AppRole auth backend role %q", path)
+	iPolicies := d.Get("policies").([]interface{})
+	policies := make([]string, 0, len(iPolicies))
+	for _, iPolicy := range iPolicies {
+		policies = append(policies, iPolicy.(string))
+	}
+
+	iCIDRs := d.Get("bound_cidr_list").([]interface{})
+	cidrs := make([]string, 0, len(iCIDRs))
+	for _, iCIDR := range iCIDRs {
+		cidrs = append(cidrs, iCIDR.(string))
+	}
+
+	data := map[string]interface{}{}
+	if v, ok := d.GetOk("period"); ok {
+		data["period"] = v.(int)
+	}
+	if len(policies) > 0 {
+		data["policies"] = policies
+	}
+	if len(cidrs) > 0 {
+		data["bound_cidr_list"] = strings.Join(cidrs, ",")
+	}
+	if v, ok := d.GetOkExists("bind_secret_id"); ok {
+		data["bind_secret_id"] = v.(bool)
+	}
+	if v, ok := d.GetOk("secret_id_num_uses"); ok {
+		data["secret_id_num_uses"] = v.(int)
+	}
+	if v, ok := d.GetOk("secret_id_ttl"); ok {
+		data["secret_id_ttl"] = v.(int)
+	}
+	if v, ok := d.GetOk("token_num_uses"); ok {
+		data["token_num_uses"] = v.(int)
+	}
+	if v, ok := d.GetOk("token_ttl"); ok {
+		data["token_ttl"] = v.(int)
+	}
+	if v, ok := d.GetOk("token_max_ttl"); ok {
+		data["token_max_ttl"] = v.(int)
+	}
+
+	_, err := client.Logical().Write(path, data)
+
+	d.SetId(path)
+
+	if err != nil {
```

Comment:
nit, not a required change: if we check `err` before we call `d.SetId`, and it's not `nil`, we can simply return the error and not worry about unsetting the ID, just set it after the error check block.

Reply:
Not 100% sure what I was thinking here. I'll update it.